File: skills/clawhub/12306-train-query/scripts/train_query.js
#!/usr/bin/env node
/**
 * 12306火车票查询脚本
 * 用法: node train_query.js [出发站] [到达站] [日期]
 * 示例: node train_query.js 长沙 北京 2026-03-16
 */

const https = require('https');
const http = require('http');
const { URL } = require('url');

// 默认查询参数
let fromStation = process.argv[2] || '长沙';
let toStation = process.argv[3] || '北京';
let date = process.argv[4] || new Date().toISOString().split('T')[0];

function getStations() {
    return new Promise((resolve, reject) => {
        const url = 'https://kyfw.12306.cn/otn/resources/js/framework/station_name.js';
        
        https.get(url, { rejectUnauthorized: false }, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
                const stations = {};
                // 格式: @bji|北京|BJP|...
                const regex = /@[^|]+\|([^|]+)\|([^|]+)\|/g;
                let match;
                while ((match = regex.exec(data)) !== null) {
                    stations[match[1]] = match[2];
                }
                resolve(stations);
            });
        }).on('error', reject);
    });
}

function queryTickets(date, fromCode, toCode) {
    return new Promise((resolve, reject) => {
        const url = `https://kyfw.12306.cn/otn/leftTicket/queryG?leftTicketDTO.train_date=${date}&leftTicketDTO.from_station=${fromCode}&leftTicketDTO.to_station=${toCode}&purpose_codes=ADULT`;
        
        const options = {
            headers: {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Safari/537.36',
                'Accept': 'application/json, text/plain, */*',
                'Accept-Language': 'zh-CN,zh;q=0.9',
                'Referer': 'https://www.12306.cn/mormhweb/',
                'Cookie': 'JSESSIONID=test123; BIGipServerotn=123456'
            },
            rejectUnauthorized: false
        };

        function fetch(url, opts, callback) {
            const client = url.startsWith('https') ? https : http;
            client.get(url, opts, (res) => {
                if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
                    const newUrl = new URL(res.headers.location, url).toString();
                    fetch(newUrl, opts, callback);
                } else {
                    let data = '';
                    res.on('data', chunk => data += chunk);
                    res.on('end', () => callback(data));
                }
            }).on('error', e => {
                console.log('Fetch Error:', e.message);
                resolve([]);
            });
        }

        fetch(url, options, (data) => {
            try {
                const json = JSON.parse(data);
                if (json.httpstatus === 200 && json.data && json.data.result) {
                    const decodedResults = json.data.result.map(item => {
                        try {
                            return decodeURIComponent(item);
                        } catch {
                            return item;
                        }
                    });
                    resolve(decodedResults);
                } else {
                    resolve([]);
                }
            } catch (e) {
                resolve([]);
            }
        });
    });
}

function parseTrain(result) {
    if (!result) return null;
    
    const items = result.split('|');
    
    const seatMap = {
        32: '商务座',
        31: '特等座', 
        30: '一等座',
        29: '二等座',
        28: '高级软卧',
        26: '软卧',
        23: '硬卧',
        21: '硬座',
        22: '无座'
    };
    
    const seats = {};
    for (const [pos, name] of Object.entries(seatMap)) {
        const idx = parseInt(pos);
        if (idx < items.length && items[idx] && items[idx] !== '无' && items[idx] !== '*') {
            seats[name] = items[idx];
        }
    }
    
    return {
        trainNo: items[3] || '',
        fromCode: items[6] || '',
        toCode: items[7] || '',
        startTime: items[8] || '--:--',
        arriveTime: items[9] || '--:--',
        duration: items[10] || '--:--',
        seats: seats
    };
}

function formatOutput(tickets, stationMap, fromName, toName) {
    if (!tickets || tickets.length === 0) {
        console.log('未查询到列车信息');
        return;
    }

    const codeToStation = {};
    for (const [name, code] of Object.entries(stationMap)) {
        codeToStation[code] = name;
    }

    console.log('\n' + '='.repeat(60));
    console.log(`🚄 ${fromName} → ${toName} 火车票查询结果`);
    console.log('='.repeat(60) + '\n');

    let count = 0;
    for (const result of tickets) {
        const info = parseTrain(result);
        if (!info || !info.trainNo) continue;

        count++;
        const from = codeToStation[info.fromCode] || info.fromCode;
        const to = codeToStation[info.toCode] || info.toCode;

        console.log(`【${info.trainNo}】${from} → ${to}`);
        console.log(`  发车: ${info.startTime}  到达: ${info.arriveTime}  历时: ${info.duration}`);
        
        if (Object.keys(info.seats).length > 0) {
            const seatStr = Object.entries(info.seats)
                .map(([name, num]) => `${name}: ${num}`)
                .join(', ');
            console.log(`  余票: ${seatStr}`);
        } else {
            console.log('  余票: 暂无');
        }
        console.log();
    }

    console.log(`共查询到 ${count} 趟列车\n`);
}

async function main() {
    try {
        console.log(`📅 查询日期: ${date}`);
        console.log(`🚉 出发站: ${fromStation}  到达站: ${toStation}`);
        console.log('-'.repeat(50));

        // 获取车站代码
        const stations = await getStations();
        const fromCode = stations[fromStation];
        const toCode = stations[toStation];

        if (!fromCode || !toCode) {
            console.log(`错误: 找不到车站代码 - ${fromStation}: ${fromCode}, ${toStation}: ${toCode}`);
            console.log('请检查车站名称是否正确');
            return;
        }

        // 查询车票
        const results = await queryTickets(date, fromCode, toCode);
        
        // 输出结果
        formatOutput(results, stations, fromStation, toStation);
        
    } catch (e) {
        console.error('Error:', e.message);
    }
}

main();

File: skills/clawhub/ly-train-12306/scripts/train_query.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
12306 高铁/火车票余票查询工具
数据源：12306 官方公开接口（kyfw.12306.cn），免费、权威、无需登录
"""
import json
import re
import sys
import os
import time
import urllib.request
import urllib.parse
import http.cookiejar

TIMEOUT = 20
BASE = "https://kyfw.12306.cn"
UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0 Safari/537.36"

# 车站代码表（首次运行自动拉取官方 station_name.js，本地缓存）
STATION_CACHE = os.path.join(os.path.dirname(os.path.abspath(__file__)), "station_codes.json")


def _fetch(url, cj, referer=None, data=None):
    """带超时的 HTTP 请求，失败立即返回 None"""
    req = urllib.request.Request(url, data=data)
    req.add_header("User-Agent", UA)
    if referer:
        req.add_header("Referer", referer)
    if data:
        req.add_header("Content-Type", "application/x-www-form-urlencoded")
    opener = urllib.request.build_opener(urllib.request.HTTPCookieProcessor(cj))
    try:
        with opener.open(req, timeout=TIMEOUT) as r:
            return r.read().decode("utf-8", errors="replace")
    except Exception:
        return None


def _load_stations():
    """加载车站代码表：优先本地缓存，否则拉取官方 station_name.js"""
    if os.path.exists(STATION_CACHE):
        try:
            with open(STATION_CACHE, encoding="utf-8") as f:
                return json.load(f)
        except Exception:
            pass
    raw = _fetch(BASE + "/otn/resources/js/framework/station_name.js?station_version=1.9300", http.cookiejar.CookieJar())
    if not raw:
        return None
    stations = {}
    for m in re.finditer(r"@([a-z]+)\|([^|]+)\|([A-Z]+)\|", raw):
        stations[m.group(2)] = m.group(3)
    if stations:
        try:
            with open(STATION_CACHE, "w", encoding="utf-8") as f:
                json.dump(stations, f, ensure_ascii=False)
        except Exception:
            pass
    return stations


def resolve_station(name, stations):
    """站名/拼音/代码模糊匹配，返回 (代码, 展示名) 列表"""
    if not stations:
        return []
    name = name.strip()
    # 1. 精确代码
    for n, c in stations.items():
        if c == name.upper():
            return [(c, n)]
    # 2. 精确站名
    if name in stations:
        return [(stations[name], name)]
    # 3. 模糊包含（站名包含关键词）
    hits = [(c, n) for n, c in stations.items() if name in n]
    if hits:
        return hits
    # 4. 拼音匹配（官方 station_name.js 第4段是拼音）
    raw = ""
    cache_file = STATION_CACHE
    if os.path.exists(cache_file):
        raw = open(cache_file, encoding="utf-8").read()
    if not raw:
        raw = _fetch(BASE + "/otn/resources/js/framework/station_name.js?station_version=1.9300", http.cookiejar.CookieJar()) or ""
    for m in re.finditer(r"@([a-z]+)\|([^|]+)\|([A-Z]+)\|([a-z]+)\|", raw):
        if m.group(4).startswith(name.lower()) or m.group(1) == name.lower():
            hits.append((m.group(3), m.group(2)))
    return hits


def query_tickets(from_name, to_name, date):
    """查询余票，返回格式化文本"""
    stations = _load_stations()
    if not stations:
        return "查询失败：无法获取车站代码表（网络异常），请稍后重试"

    from_hits = resolve_station(from_name, stations)
    to_hits = resolve_station(to_name, stations)
    if not from_hits:
        return "未找到车站：" + from_name + "，请检查站名（如：深圳北、邯郸东）"
    if not to_hits:
        return "未找到车站：" + to_name + "，请检查站名（如：深圳北、邯郸东）"

    # 多个匹配时按名称排序取第一个（如"深圳"→深圳站），提示用户可用更精确站名
    from_code, from_label = from_hits[0]
    to_code, to_label = to_hits[0]
    hint = ""
    if len(from_hits) > 1:
        hint = "（" + from_name + " 匹配多站：" + ", ".join(n for c, n in from_hits) + "，默认取 " + from_label + "）"
    if len(to_hits) > 1:
        hint += "（" + to_name + " 匹配多站：" + ", ".join(n for c, n in to_hits) + "，默认取 " + to_label + "）"

    # 建立 cookie 会话（12306 需要先访问 init 页面拿 cookie）
    cj = http.cookiejar.CookieJar()
    _fetch(BASE + "/otn/leftTicket/init?linktypeid=dc", cj, referer=BASE + "/otn/index/init")

    # 代码→站名反向映射（显示用）
    code2name = {c: n for n, c in stations.items()}

    params = urllib.parse.urlencode({
        "leftTicketDTO.train_date": date,
        "leftTicketDTO.from_station": from_code,
        "leftTicketDTO.to_station": to_code,
        "purpose_codes": "ADULT",
    })
    url = BASE + "/otn/leftTicket/queryG?" + params
    raw = _fetch(url, cj, referer=BASE + "/otn/leftTicket/init")
    if raw is None:
        return "查询失败：请求 12306 超时，请稍后重试"
    if raw.lstrip().startswith("<!DOCTYPE") or "<html" in raw[:200]:
        return (f"查询失败：{date} 可能超出预售期（12306 一般提前 15 天放票）或请求被限流，请换个日期或稍后重试")
    try:
        data = json.loads(raw)
    except Exception:
        return "查询失败：12306 返回异常，请稍后重试"

    if not data.get("status"):
        return "查询失败：12306 接口未返回数据（可能被限流），请稍后重试"

    result = data.get("data", {}).get("result", [])
    if not result:
        return f"未查询到 {from_label} -> {to_label} {date} 的车次（可能日期超出预售期或当日无车）"

    lines = [f"🚄 {from_label} → {to_label}（{date}）共 {len(result)} 个车次：" + (hint or ""), ""]
    seat_headers = [("商务座", 32), ("特等座", 20), ("一等座", 31), ("二等座", 30),
                    ("软卧", 23), ("动卧", 27), ("硬卧", 29), ("硬座", 26), ("无座", 22)]

    for i, item in enumerate(result, 1):
        p = item.split("|")
        if len(p) < 33:
            continue
        no = p[3]
        dep_code, dep_t = p[6], p[8]
        arr_code, arr_t = p[7], p[9]
        dur = p[10]
        can_book = p[11]

        dep_name = code2name.get(dep_code, dep_code)
        arr_name = code2name.get(arr_code, arr_code)

        seats = []
        for label, idx in seat_headers:
            if idx < len(p):
                v = p[idx]
                if v in ("", "无"):
                    continue
                if v == "*":
                    v = "候补"
                seats.append(f"{label}:{v}")
        seat_str = " | ".join(seats) if seats else "暂无余票"

        book_flag = "✅可订" if can_book == "Y" else "❌不可订"
        lines.append(f"{i}. {no} {dep_name}{dep_t} → {arr_name}{arr_t}（历时{dur}）{book_flag}")
        lines.append(f"   余票：{seat_str}")
        lines.append("")

    lines.append("数据来源：12306 官方（kyfw.12306.cn）｜余票实时变动，以 12306 为准")
    return "\n".join(lines)


def main():
    args = sys.argv[1:]
    if len(args) < 3:
        print("用法：python3 train_query.py <出发站> <到达站> <日期YYYY-MM-DD>")
        print("示例：python3 train_query.py 深圳北 邯郸东 2026-09-03")
        print("       python3 train_query.py 深圳 邯郸 2026-09-03,2026-09-05 （多日期）")
        return
    frm, to = args[0], args[1]
    dates = args[2].split(",")
    for i, d in enumerate(dates):
        if i > 0:
            print("")
        print(query_tickets(frm, to, d.strip()))


if __name__ == "__main__":
    main()
